### ISSUE TITLE
Explanation of the way of form handling

### ISSUE THREAD

**Issue by nikiforosbotis** [TEAMMATES/teammates#7209]
In the "instructorCourseEnroll.jsp" file, there is a form and the action button directs the users here: action="${data.instructorCourseEnrollSaveLink}.
![1](https://cloud.githubusercontent.com/assets/9116253/25556710/0433aa4e-2d0a-11e7-94f3-b25132db23b5.png)

I was wondering where this "method" (instructorCourseEnrollSaveLink) is located because I want to check how the method's data is handled. I did try to search it in GitHub but I could not find any relevant detail.
![2](https://cloud.githubusercontent.com/assets/9116253/25556708/fe085232-2d09-11e7-9e1c-96dd5f7e118a.png)

**Comment by VamsiSangam** [comment]
Hi @nikiforosbotis ! The method you are looking for is the `execute()` method in `teammates.ui.controller.InstructorCourseEnrollAction.java`.

You can learn a lot about the architecture and design of the TEAMMATES application in [design.md](https://github.com/TEAMMATES/teammates/blob/master/docs/design.md). Although even I'm not yet fully aware of the complete architecture of the application, it is mentioned in the docs that -

> [logic] - [ui::view] - [ui::controller] represent an application of Model-View-Controller pattern. 

In MVC pattern, the requests are handled by the controller. So if you want to search for a method which is handling a request, then you can blindly start looking for it in the `teammates.ui.controller` package. :smiley:

**Comment by AnaghHegde** [comment]
Hi @nikiforosbotis
>I was wondering where this "method" (instructorCourseEnrollSaveLink) is located because I want to check how the method's data is handled.

 I am no perfect even i had a similar problem. But i know that those methods will be there in `instructor*.es6` files  but like @VamsiSangam said then those methods will make use of the backend java files. For more information take a look at the  [design.md](https://github.com/TEAMMATES/teammates/blob/master/docs/design.md) as @VamsiSangam said. But for your question may be [design.md-user-inoked-requests](https://github.com/TEAMMATES/teammates/blob/master/docs/design.md#user-invoked-requests) will help you.

**Comment by nikiforosbotis** [comment]
@VamsiSangam at first, thank you for your answer! I understand your point but I would like to ask you:
1)Do you mean the teammates.ui.controller.InstructorCourseEnrollPageAction.java file? because I cannot find the exact filename that you sent me.
2)So there is no such method with the exact name: "instructorCourseEnrollSaveLink" right? I mean I cannot understand how the action: ${data.instructorCourseEnrollSaveLink} is "transformed" to the execute() method, as you are saying.

Thank you!

**Comment by AnaghHegde** [comment]
@nikiforosbotis 
So there is no such method with the exact name: "instructorCourseEnrollSaveLink" right?

Did you read the [answer](https://github.com/TEAMMATES/teammates/issues/7209#issuecomment-298209976) i have written above. I think i have mentioned that part. 

> But i know that those methods will be there in instructor*.es6

**Comment by nikiforosbotis** [comment]
@AnaghHegde Thanks for your help.

The only two files with a similar name (but not the same) are these, as shown in the TEAMMATES directory:
I guess I can find the functionality I want by checking out these files. 

![1](https://cloud.githubusercontent.com/assets/9116253/25563742/2e6129d0-2dab-11e7-8eb5-bc5aa14c420b.png)

**Comment by AnaghHegde** [comment]
@nikiforosbotis Did you get answer to your question ?. I mean is everything clear now ?.

**Comment by VamsiSangam** [comment]
> 1)Do you mean the teammates.ui.controller.InstructorCourseEnrollPageAction.java file? because I cannot find the exact filename that you sent me.

Very sorry, my bad! I meant `teammates.ui.controller.InstructorCourseEnrollSaveAction.java`.

> I mean I cannot understand how the action: ${data.instructorCourseEnrollSaveLink} is "transformed" to the execute() method, as you are saying.

I am going to give a little detailed explanation, please bear with me. So when you send a request to `/page/instructorCourseEnrollPage`, according to the `web.xml` code -

``` xml
<servlet>
        <description>Servlet that handles all incoming requests</description>
        <servlet-name>ControllerServlet</servlet-name>
        <servlet-class>teammates.ui.controller.ControllerServlet</servlet-class>
</servlet>
```

The `ControllerServlet` class handles every request. There, the `doPost()` method is called, where we have -

``` java
ActionResult actionResult = c.executeAndPostProcess();
actionResult.send(req, resp);
```

Here, the `executeAndPostProcess()` causes the `execute()` method in `InstructorCourseEnrollPageAction` to run. If you have noticed in the `execute()` method, you create an object named `pageData` which is of the type `InstructorCourseEnrollPageData`. `InstructorCourseEnrollPageData` has a method `getInstructorCourseEnrollSaveLink()`, it is this method which constructs the end output. Eg. -> `/page/instructorCourseEnrollSave?courseid=teammates.instructor.uni-demo&user=teammates.instructor%40university.edu`

The `execute()` method returns a `ShowPageResult` object -

``` java
return createShowPageResult(Const.ViewURIs.INSTRUCTOR_COURSE_ENROLL, pageData);
```

So if you go back to the code above -

``` java
ActionResult actionResult = c.executeAndPostProcess(); // actionResult is a ShowPageResult object
actionResult.send(req, resp); // this calls send() method in ShowPageResult class
```

So in `ShowPageResult`'s `send()` method, you have a super magical line -

``` java
req.setAttribute("data", data);
```

Where the `data` object here is of type `InstructorCourseEnrollPageData`. That line is the reason you are able to write `${data}` in your `.jsp` file. When you use `setAttribute` in your java code, you access it via `${}` in JSP. Now, `data` is of the type `InstructorCourseEnrollPageData`, that is why you are able to write `${data.instructorCourseEnrollSaveLink}` which actually calls that method I mentioned earlier, giving you the required URL with required parameters.

---
**In a nutshell** - "data" in `${data.instructorCourseEnrollSaveLink}` is an object of type `InstructorCourseEnrollPageData` which has a property `instructorCourseEnrollSaveLink`. Writing `${data.instructorCourseEnrollSaveLink}` invokes `InstructorCourseEnrollPageData.getInstructorCourseEnrollSaveLink()`

Hope this clears your question :smile:

**Comment by AnaghHegde** [comment]
@VamsiSangam That is well explained. I hope it clarifies @nikiforosbotis doubt. On a note @nikiforosbotis feel free to ask if you have still have any doubts. We are happy to help.

**Comment by nikiforosbotis** [comment]
@VamsiSangam your answer is very detailed and it helped me a lot! thank you very much for that. Thank you also  @AnaghHegde for your help. If I have another question on that topic, I will let you know.

**Comment by nikiforosbotis** [comment]
Guys @VamsiSangam , @AnaghHegde I did check your very detailed answers and I understood many things considering how the TEAMMATES is working.

**One more question please:** 

On the getInstructorCourseEnrollSaveLink(String courseId) method of the PageData.java file, there is a link for the INSTRUCTOR_COURSE_ENROLL_SAVE page that is translated to: "/page/instructorCourseEnrollSave"; according to the Const.java file. 

This is confirmed by the page that is shown after the form is submitted:(/page/instructorCourseEnrollSave?courseid=CS123&user=test%40example.com):

![instructorcourseenrollsave](https://cloud.githubusercontent.com/assets/9116253/25582773/949d2900-2e97-11e7-8138-b0828c36604f.png)

However, I searched in GitHub for the page: `page/instructorCourseEnrollSave` and I did not find anything similar. What I did find was this jsp page: `/jsp/instructorCourseEnrollResult.jsp` which seems to be the same as the required. **Is there a way which "transforms" the first page to the .jsp one? Otherwise, where is the `page/instructorCourseEnrollSave` located?**

Do you get what I mean?

Thank you in advance.

**Comment by VamsiSangam** [comment]
Hi @nikiforosbotis ! So as things start with `ControllerServlet.java`'s `doPost()` method, there we have -

``` java
Action c = new ActionFactory().getAction(req);
```

So, the purpose of the `getAction()` method is to return the proper `Action` subclass associated with the request. So this is the `getAction()` which gets called -

``` java
public Action getAction(HttpServletRequest req) {
     String url = req.getRequestURL().toString();
     log.info("URL received : [" + req.getMethod() + "] " + url);

     String uri = req.getRequestURI();
     if (uri.contains(";")) {
         uri = uri.split(";")[0];
     }
     Action c = getAction(uri);
     c.init(req);
     return c;

 }
```

Here, in the line `Action c = getAction(uri);`, we have passed the URL to get the proper `Action` subclass. Inside the `getAction(uri)` method there's a line -

``` java
Class<? extends Action> controllerClass = actionMappings.get(uri);
```

It is using a variable `actionMappings` to get the proper `Action` subclass associated with the requested URL. Now if you look at the top of `ActionFactory.java` file, you will find the declaration of `actionMappings` which is a `HashMap<String, Class<? extends Action>>()` and just below it is a huuuuuugely important code in the static block which initializes the `actionMappings` variable. So there you have a line -

``` java
map(INSTRUCTOR_COURSE_ENROLL_SAVE, InstructorCourseEnrollSaveAction.class);
```

This line associates `/page/instructorCourseEnrollSave` to `InstructorCourseEnrollSaveAction.java`. So that's the first half of your answer.

So due to the association above `InstructorCourseEnrollSaveAction`'s `execute()` method gets called (similar to my previous answer as to how `InstructorCourseEnrollPageAction`'s `execute()` gets called).

In `InstructorCourseEnrollSaveAction`'s `execute()` method you have the line -

``` java
return createShowPageResult(Const.ViewURIs.INSTRUCTOR_COURSE_ENROLL_RESULT, pageData);
```

that line associates `/jsp/instructorCourseEnrollResult.jsp` with the page data (data which is displayed) and that's the view which is rendered.

So, basically `/page/instructorCourseEnrollSave` is associated with `InstructorCourseEnrollSaveAction.java` which returns the view `/jsp/instructorCourseEnrollResult.jsp`.

Think about it for a while. I'm sure you'll get it! Let me know if this answers your question. :smile:

**Comment by nikiforosbotis** [comment]
@VamsiSangam Thank you very much for your detailed response. It really helped me a lot.
